### PR TITLE
[RLlib] Smaller eval worker set fixes.

### PR DESCRIPTION
**Pull request by sven1977**
Signed-off-by: sven1977 <svenmika1977@gmail.com>

Smaller eval worker set fixes.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm.py [hunk 1]
```diff
@@ -1147,6 +1148,11 @@ def remote_fn(worker, w_ref, w_seq_no):
         # subsequent step results as latest evaluation result.
         self.evaluation_metrics = {"evaluation": metrics}
 
+        # Trigger `on_evaluate_end` callback.
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:1151]: This callback was missing for `enable_async_evaluation=True` setting.

File: rllib/algorithms/algorithm.py [hunk 2]
```diff
@@ -2701,7 +2707,11 @@ def _run_one_evaluation(
             if self.evaluation_workers is not None
             else 0
         )
-        eval_results["evaluation"]["num_recreated_workers"] = num_recreated
+        # Worker failures might have already been handled within `self._evaluate_async`
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2710]: This count would be overridden by 0 for `enable_async_evaluation=True`.

Comment (gjoliver) [reply rllib/algorithms/algorithm.py:2710]: 🤔 wonder if we should move this try_recover_from_step_attempt() into self.evaluate() then.
or, I actually need to refactor all the remote gets anyways for elastic training. so maybe I can clean this up when I get to remote_req_manager.

**Review by gjoliver — APPROVED** [review]
ok, can you maybe add a TODO for me above the if statement, like:
```
TODO(jungong) : revisit after elastic async evaluation is done.
```

**Pull request merged**